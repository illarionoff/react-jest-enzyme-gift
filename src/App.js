import React, { Component } from "react";
import "./App.css";
import Gift from "./components/Gift/Gift";

class App extends Component {
  state = {
    gifts: []
  };

  addGift = () => {
    const ids = this.state.gifts.map(gift => gift.id);
    const max_id = ids.length > 0 ? Math.max(...ids) : 0;
    this.setState({
      gifts: [...this.state.gifts, { id: max_id + 1 }]
    });
  };

  removeGift = id => {
    this.setState({
      gifts: this.state.gifts.filter(gift => gift.id !== id)
    });
  };

  render() {
    return (
      <div className="App">
        <h2>Hello</h2>
        <button className="btn-add" onClick={this.addGift}>
          Add
        </button>
        <div className="gift-list">
          {this.state.gifts.map(gift => {
            return (
              <Gift key={gift.id} gift={gift} removeGift={this.removeGift} />
            );
          })}
        </div>
      </div>
    );
  }
}

export default App;
