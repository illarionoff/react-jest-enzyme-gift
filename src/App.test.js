import React from "react";
import { shallow } from "enzyme";
import App from "./App";
import toJson from "enzyme-to-json";

describe("renders App component", () => {
  const wrapper = shallow(<App />);

  test("renders without error", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test("initializes App with empty 'gifts' state", () => {
    expect(wrapper.state().gifts).toEqual([]);
  });

  describe("when clicking 'add-gift' button", () => {
    const id = 1;

    beforeEach(() => {
      wrapper.find(".btn-add").simulate("click");
    });

    afterEach(() => {
      wrapper.setState({ gifts: [] });
    });

    test("adds new 'gift' to 'gifts' state", () => {
      expect(wrapper.state().gifts).toEqual([{ id }]);
    });

    test("adds new div 'gift-item'", () => {
      expect(wrapper.find(".gift-list").children().length).toEqual(1);
    });

    test("adds new div 'gift-item'", () => {
      expect(wrapper.find("Gift").exists()).toBe(true);
    });

    describe("when user clicks 'delete' button", () => {
      beforeEach(() => {
        wrapper.instance().removeGift(id);
      });

      test("removes the gift from state 'state' by id", () => {
        expect(wrapper.state().gifts).toEqual([]);
      });
    });
  });
});
