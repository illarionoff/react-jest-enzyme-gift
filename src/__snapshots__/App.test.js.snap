// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component renders without error 1`] = `
<div
  className="App"
>
  <h2>
    Hello
  </h2>
  <button
    className="btn-add"
    onClick={[Function]}
  >
    Add
  </button>
  <div
    className="gift-list"
  />
</div>
`;
